---
title: "Process Stacks Output"
author: "Amanda Stahlke"
date: "2/26/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r libs}
library(dplyr)
library(ggplot2)
```

# Global dataset

Produced in Stacks/2.5 with script raw2vcf.sh. Briefly:
``````{r, engine = 'bash', eval = FALSE, stackscode}
projhome=/mnt/ceph/stah3621/diorhabda/active_stacks/StahlkeBitume_etal_2019

#raw1=/mnt/ceph/stah3621/diorhabda/00rawdata/RADseq/widesurvey ## produced in 2016ish
#raw2=/mnt/ceph/stah3621/diorhabda/00rawdata/RADseq/morewidesurvey ## produced later
## Combined here ##
raw=/mnt/ceph/stah3621/diorhabda/00rawdata/RADseq/StahlkeBitume

fqc_out=$projhome/fastqc
cf_out=$projhome/1clonefilter ## output directory
proc=$projhome/2processed ## these are your clone_filtered, demultiplexed reads
al_out=$projhome/3refaligned ## output directory
popmap=${projhome}/info/masterpopmap.tsv
g_out=$projhome/4gstacks
p_out=$projhome/5populations

for lib in $(ls ${raw}/PH*R1*); do
       echo $lib
       base=$(basename $lib _L00*_R1_001.fastq.gz)
#       echo $base
                ##T10_S17_L002_R
       T=$(echo $base | cut -f 1 -d _)
       echo $T

        ### clone_filter
       clone_filter -P -1 $lib -2 $raw/$T*R2_001.fastq.gz -o $cf_out -i gzfastq \
               &>> $cf_out/clone_filter_$T.log
$cf_out/clone_filter_$T.log

       process_radtags -1 $cf_out/${T}*_R1_001.1.fq.gz \
       -2 $cf_out/${T}*_R2_001.2.fq.gz --paired --bestrad -e sbfI -r -q \
       -b $projhome/info/${T}_barcodes.tsv -o $proc --barcode_dist_1 3
       mv $proc/process_radtags.1clonefilter.log $proc/process_radtags.1clonefilter.${T}.log
       mv $proc/*rem*fq.gz $proc/rem/
$proc/process_radtags.1clonefilter.${T}.log
       done

#### Align fq to reference ####
bowtie_db=/mnt/ceph/stah3621/diorhabda/bowtie_db/carinulata_arced_scaffolds
bowtielog=$al_out/bowtie.log

module load bowtie2
module load samtools

for samp in $(ls $proc/*1.fq.gz | cut -f 9 -d / | cut -f 1 -d .); do
        echo 'Aligning' $samp >> $bowtielog

        ## I use --very-sensitive because divergent spp
        bowtie2 --very-sensitive -x $bowtie_db \
       -1 $proc/${samp}.1.fq.gz -2 $proc/${samp}.2.fq.gz \
        -S $al_out/${samp}.sam 2>> $bowtielog
         samtools view -bS $al_out/${samp}.sam | samtools sort - --threads 30 \
         -o $al_out/${samp}.bam
done

gstacks -I $al_out/ -M $popmap -O $g_out/ -t 30

stacks-dist-extract $g_out/gstacks.log.distribs bam_stats_per_sample > $g_out/bam_stats_per_sample.tsv

stacks-dist-extract $g_out/gstacks.log.distribs effective_coverages_per_sample > $g_out/effective_coverages_per_sample.tsv

populations -P $g_out -O $p_out -M $popmap --threads 30 -e SbfI --merge-sites \
         --write-random-snp --ordered-export --vcf

module load vcftools

## individual missingness - .imiss
vcftools --missing-indv --vcf $p_out/populations.snps.vcf \
         --out $f_out/populations.snps

## missingness across sites - .lmiss
vcftools --missing-site --vcf $p_out/populations.snps.vcf \
         --out $f_out/populations.snps

mkdir R50
### winner - reasonable number of loci, snps, and missing data rate for gloabl analysis

populations -P ${g_out} -O ${p_out}/R50 -M ${popmap} --threads 30 \
-e SbfI --merge-sites --write-random-snp --ordered-export --vcf \
         -R 0.50

vcftools --missing-indv --vcf $p_out/populations.snps.vcf \
         --out $f_out/populations.snps

mkdir ${p_out}/75mmR50
awk '$5 > 0.75' ${p_out}/R50/populations.snps.imiss | cut -f 1 > ${p_out}/75mmR50/75mmR50.txt
grep -v -f ${p_out}/75mmR50/75mmR50.txt $popmap > ${p_out}/75mmR50/75mmR50_popmap.txt

populations -P ${g_out} -O ${p_out}/75mmR50 -M ${p_out}/75mmR50/75mmR50_popmap.txt \
         --threads 30 \
         -e SbfI --merge-sites \
         --write-random-snp --ordered-export --vcf --structure \
         -R 0.50

```

Final sample and site list.
```{r}
global_popmap <- read.delim("../info/globalpopmap_R5075mm4cov.tsv",
           col.names=c("sampID", "popID"),
           header = F)
table(global_popmap$popID)
```

Merge this with existing metadata table. 

## Samtools flagstat
Report generated from Multiqc
```{r flagstat}
flagstat <- read.delim("../3refaligned/multiqc_data/multiqc_samtools_flagstat.txt")
head(flagstat)

mapped_perc <- select(flagstat, c(Sample,mapped_passed_pct))
hist(mapped_perc$mapped_passed_pct, breaks = 50)
mean(mapped_perc$mapped_passed_pct)

colnames(mapped_perc)[1] <- "sampID"

```

```{r violinplot}
sample_meta.dat <- read.csv("../info/pop_metadata.csv", header = TRUE, stringsAsFactors = FALSE)
sample_meta.dat

mapped_perc_metadat <- merge(mapped_perc, global_popmap) %>%
  merge(sample_meta.dat)

ggplot(mapped_perc_metadat) +
  geom_violin(aes(x = Grp, y = mapped_passed_pct)) +
  geom_point(aes(x = Grp, y = mapped_passed_pct), alpha = .5) +
  geom_hline(yintercept  = as.numeric(summarise(mapped_perc_metadat, mean(mapped_passed_pct)))) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
ggsave("mapped_perc_metadat_grp.jpg", dpi = 300)

```


## Gstacks

```{r gstacks_bamstats}
bam <- read.delim(file = "../3refaligned/bam_stats_per_sample.tsv")
hist(bam$records, breaks=1000)
hist(bam$records, breaks=1000, xlim = c(0,1000000))
bam %>%
  dplyr::filter(records < 1000)
```



```{r gstacks_effcov}
eff_cov <- read.delim("../4gstacks/effective_coverages_per_sample.tsv", skip = 2)
hist(eff_cov$mean_cov, breaks = 100, xlim = c(5, 50))

## check which rep to keep among these, kept the higher coverage
eff_cov[grep("_", eff_cov$sample),]

colnames(global_popmap)[1] <- "sample"

## No samples here with coverage less than 4
merge(eff_cov, global_popmap) %>%
    # merge(global_popmap) %>%
  dplyr::filter(mean_cov_ns < 4) %>%
    arrange(mean_cov_ns) %>%
  select(sample)
  # %>% write.table(., file = "low_cov_samples.tsv", quote = F, col.names = F, row.names = F)

merge(eff_cov, global_popmap) %>%
  group_by(popID) %>%
  summarise(mean(mean_cov))

merge(eff_cov, global_popmap) %>%
  merge(sample_meta.dat) %>%
  group_by(Grp) %>%
  summarise(mean(mean_cov))

merge(eff_cov, global_popmap) %>%
  # group_by(popID) %>%
  summarise(mean(mean_cov))

colnames(sample_meta.dat)
ggstatsplot::ggbetweenstats(data= merge(eff_cov, global_popmap) %>%
                              merge(sample_meta.dat), 
  # group_by(popID),
  pairwise.comparisons = TRUE,
                            x = Grp,
                            y = mean_cov)

```

  merge(popmap)## Missing data
```{r ind_missingness}
imiss <- read.delim("../5populations/R50/populations.snps.imiss")
head(imiss)
hist(imiss$F_MISS, breaks = 100)

colnames(imiss)[1] <- "sampID"
colnames(global_popmap)[1] <- "sampID"

merge(imiss, global_popmap) %>%
    arrange(desc(F_MISS)) 

miss_meta <- merge(imiss, global_popmap) %>%
  merge(sample_meta.dat)

arrange(miss_meta, desc(F_MISS)) 

summarise(miss_meta, mean(F_MISS)) 

miss_meta %>%
  group_by(Grp) %>%
  summarise(mean= mean(F_MISS))

ggplot(miss_meta) +
  geom_violin(aes(x = Grp, y = F_MISS)) +
  geom_point(aes(x = Grp, y = F_MISS), alpha = .5) +
  geom_hline(yintercept  = as.numeric(summarise(miss_meta, mean(F_MISS)))) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) 
# ggsave("missingdata_metadat_grp.jpg", dpi = 300, width = 6, height = 4, units = "in")

ggstatsplot::ggbetweenstats(data= miss_meta, 
                            x = Grp,
                            y = F_MISS, 
                            outlier.label = popID)
```
